feat: support `shape_touched` from Dask

agoose77 opened the pull request:
This PR *at present* represents a hack of the existing uproot remapping loader to support dask-awkward's new buffer projection interface.

@lgray the Coffea `form_mapping` object that is passed to `uproot.dask` needs to satisfy the `ImplementsFormMapping` interface:
```python
class ImplementsFormMapping(Protocol):
    def __call__(self, form: Form) -> tuple[Form, ImplementsFormMappingInfo]:
        ...
```

The idea is that the form mapping generates _both_ the remapped form _and_ any auxiliary state required for projection. The return tuple's second item, `ImplementsFormMappingInfo` has the following protocol:
```python
class ImplementsFormMappingInfo(Protocol):
    @property
    def behavior(self) -> dict | None:
        ...

    buffer_key: str | Callable

    def parse_buffer_key(self, buffer_key: str) -> tuple[str, str]:
        ...

    def keys_for_buffer_keys(self, buffer_keys: frozenset[str]) -> frozenset[str]:
        ...

    def load_buffers(
        self,
        tree: HasBranches,
        keys: frozenset[str],
        start: int,
        stop: int,
        options: Any,
    ) -> Mapping[str, AwkArray]:
        ...
```

This PR hacks the `_map_schema_uproot` object to provide this interface, but it could be cleaner; one could remove any unused code here.

agoose77 committed:
wip: initial commit

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

agoose77 committed:
fix: rename function

agoose77 committed:
Merge remote-tracking branch 'agoose77/agoose77/feat-support-shape-touched' into agoose77/feat-support-shape-touched

agoose77 committed:
fix: use report_necessary_buffers

agoose77 committed:
Merge branch 'master' into agoose77/feat-support-shape-touched

agoose77 commented on src/coffea/processor/executor.py:
```diff
@@ -1718,7 +1718,7 @@ def _work_function(
 
                     to_compute = processor_instance.process(events)
-                    materialized = dask_awkward.necessary_columns(to_compute)
+                    # materialized = dask_awkward.report_necessary_buffers(to_compute)
```
@lgray I wasn't clear on the scope of `materialized`, as it's populated in two different places. Could you figure out what you'd like to do here? `dak.report_necessary_buffers` produces a mapping from layer name to a `NecessaryBuffers` or `None`. The `NecessaryBuffers` namedtuple contains two fields, one for data+shape touched buffers, the other for shape-only.

lgray replied:
Oh this code is about to be removed in any case that was kept around to keep tests passing and exercise processing workflows.

It's doing the job of some old code in awkward1, the pattern for which is now obsolete.

agoose77 committed:
fix: properly parse form keys

agoose77 committed:
Merge remote-tracking branch 'agoose77/agoose77/feat-support-shape-touched' into agoose77/feat-support-shape-touched

agoose77 committed:
hack: convert Content to array

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

agoose77 committed:
fix: ensure layout nodes converted to arrays

agoose77 committed:
Merge remote-tracking branch 'agoose77/agoose77/feat-support-shape-touched' into agoose77/feat-support-shape-touched

lgray committed:
adjust coffea pins to latest releases and pre-releases

agoose77 commented on pyproject.toml:
```diff
@@ -37,11 +37,11 @@ classifiers = [
   "dask[array]>=2023.4.0",
-  "dask-awkward>=2023.7.1,!=2023.8.0",
-  "dask-histogram>=2023.6.0",
+  "dask-awkward>=2023.10a1,!=2023.8.0",
```
```suggestion
  "dask-awkward>=2023.10a1",
```

lgray committed:
use pytorch-only triton image

lgray committed:
streamline version requirements

Co-authored-by: Angus Hollands <goosey15@gmail.com>

lgray committed:
Merge branch 'master' into agoose77/feat-support-shape-touched

lgray committed:
Merge branch 'master' into agoose77/feat-support-shape-touched

agoose77 committed:
fix: don't import protocol

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lgray committed:
Merge branch 'master' into agoose77/feat-support-shape-touched

lgray committed:
remove deprecated interface definition

agoose77 commented on src/coffea/nanoevents/factory.py:
```diff
@@ -125,7 +141,41 @@ def __call__(self, form):
+
+        return TranslateBufferKeys()
 
     def create_column_mapping_and_key(self, tree, start, stop, interp_options):
```
@lgray can we remove this from the base definition, or does someone out there likely use it?

lgray replied:
interface is completely private and rather obscure to boot, so I am fine to remove it.

agoose77 commented on tests/test_jetmet_tools.py:
```diff
@@ -837,9 +837,9 @@ def test_corrected_jets_factory(optimization_enabled):
         )
 
-        print(dak.necessary_columns(jets.eta))
+        print(dak.report_necessary_buffers(jets.eta))
```
```suggestion
        print(dak.report_necessary_columns(jets.eta))
```

lgray replied:
is `dak.necessary_columns` deprecated? if so on what time scale?

agoose77 replied:
`necessary_columns` → `report_necessary_columns`.

We can make it as long as necessary. At one point, I was going to remove it in favour of `report_necessary_buffers`, but the new `report_necessary_columns` is designed to allow IO layers to report columns iff. they support them.

agoose77 commented on tests/test_jetmet_tools.py:
```diff
@@ -837,9 +837,9 @@ def test_corrected_jets_factory(optimization_enabled):
+        print(dak.report_necessary_buffers(jets.eta))
         print(
-            dak.necessary_columns(
+            dak.report_necessary_buffers(
```
```suggestion
            dak.report_necessary_columns(
```

agoose77 committed:
Update tests/test_jetmet_tools.py

agoose77 committed:
Update tests/test_jetmet_tools.py

agoose77 commented on tests/test_jetmet_tools.py:
```diff
@@ -837,9 +837,9 @@ def test_corrected_jets_factory(optimization_enabled):
         )
 
-        print(dak.necessary_columns(jets.eta))
+        print(dak.report_necessary_columns(jets.eta))
```
@lgray note that `report_necessary_columns` reports whatever the IO layer thinks a "column" is. It doesn't necessarily mean "dotted field" right now, rather I envisage it talking about the concept of a "key" or "column" that can be read from the IO source. 

@douglasdavis this is something I didn't fully think about when we made `necessary_columns` shim `report_necessary_columns`.

The difference will be that for remapped forms, the `report_necessary_columns` (and by proxy, `necessary_columns`) will report the underlying ROOT file keys, not the remapped keys.

agoose77 replied:
It therefore might mean that we should remove that alias for `necessary_columns`.

lgray replied:
I actually prefer the new behavior since users will probably expect to get a list of columns in the file they're reading! But seeing the remapped keys is nice for debugging.

lgray replied:
Can we make it so that we get a version or two of dask-awkward before `dak.necessary_columns` is gone and give migration instructions? I know a few end-users that use that function.

agoose77 replied:
@lgray I'm not 100% clear if we're on the same page. 

`report_necessary_columns` is a new function, so I'm comfortable with its semantics deviating from `ak.necessary_columns`. 

It would be possible to restore the old behavior, but it will require a reasonable amount of work on the uproot side (I am guessing).

`ak.necessary_columns` is currently an alias for this new function, but for uproot this wouldn't be a non-breaking change; if previously it would report `ROOT_KEY.record.foo.bar`, now it would report `ROOT_KEY`. Is this OK? I don't see it being that useful if it breaks existing users.

lgray replied:
OK I suppose for my most typical use case I never see keys with `ROOT_KEY.record.foo.bar` formatting, it's really only ever `ROOT_KEY` and any structure is up in the form being remapped to. So they appear the same to me.

lgray committed:
remove further remnants of old remapping interface

agoose77 committed:
refactor: make key translation obvious

agoose77 commented on src/coffea/nanoevents/factory.py:
```diff
@@ -77,24 +76,50 @@ def __init__(
+            operands = urllib.parse.unquote(form_key).split(",")
+
+            it_operands = iter(operands)
+            next(it_operands)
```
This function behaves like `extract_form_key_base_columns`, but supports operands e.g. attribute lookup.

agoose77 commented on src/coffea/nanoevents/factory.py:
```diff
@@ -147,8 +175,17 @@ def create_column_mapping_and_key(self, tree, start, stop, interp_options):
             use_ak_forth=True,
         )
         mapping.preload_column_source(partition_key[0], partition_key[1], tree)
```
I've deliberately tried not to touch this logic at all. Hopefully, by presenting the correct interface to uproot, internal refactoring can happen here in future without any side effects.

lgray committed:
fix typo from refactor

lgray committed:
update pins (note uncapped numpy and numba skooch)

lgray committed:
try to convince pip to upgrade numpy upon installing coffea

lgray committed:
be more insistent

lgray committed:
numba 0.58 pins numpy from above < 1.26

nb: safer to repin in coffea for users because of numba's sliding window, very easy to get a mismatch

nsmith- commented on src/coffea/processor/executor.py:
```diff
@@ -1734,11 +1734,11 @@ def _work_function(
                     if isinstance(file, uproot.ReadOnlyDirectory):
                         metrics["bytesread"] = file.file.source.num_requested_bytes
+                    # metrics["data_and_shape_buffers"] = set(materialized)
+                    # metrics["shape_only_buffers"] = set(materialized)
```
Are these placeholders for something in the future?

lgray replied:
This is just to get tests to run - executor is going to be significantly changed in a subsequent PR.

nsmith- commented on src/coffea/nanoevents/factory.py:
```diff
@@ -77,24 +76,50 @@ def __init__(
+    def parse_buffer_key(self, buffer_key):
+        prefix, attribute, form_key = buffer_key.rsplit("/", maxsplit=2)
+        if attribute == "offsets":
+            return (form_key[: -len("%2C%21offsets")], attribute)
```
This seems to also be in the previous code as well, but some of these string splitting and urllib name mangling things were put into https://github.com/CoffeaTeam/coffea/blob/master/src/coffea/nanoevents/util.py and could be used here

nsmith- commented on src/coffea/nanoevents/factory.py:
```diff
@@ -68,7 +67,7 @@ def _key_formatter(prefix, form_key, form, attribute):
 
 
-class _map_schema_base(ImplementsFormTransformation):
+class _map_schema_base:  # ImplementsFormMapping, ImplementsFormMappingInfo
```
Any reason not to inherit the protocol(s)?

agoose77 replied:
They weren't exported yet! I think a new release includes them. Let me check.

nsmith- reviewed (approved):
We need to do some cleanup, but it can happen later I think.
@agoose77 does the protocol you defined in the initial comment have some explanatory docstrings available somewhere?

lgray committed:
clean up usage of quoted ",!offsets"

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lgray committed:
flake8 lint
